File: ontrack-extension-api/src/test/java/net/nemerosa/ontrack/extension/api/support/TestDecoratorPropertyType.java
package net.nemerosa.ontrack.extension.api.support;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.extension.support.AbstractPropertyType;
import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.Text;
import net.nemerosa.ontrack.model.form.YesNo;
import net.nemerosa.ontrack.model.security.ProjectEdit;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;

@Component
public class TestDecoratorPropertyType extends AbstractPropertyType<TestDecorationData> {

    @Autowired
    public TestDecoratorPropertyType(TestExtensionFeature extensionFeature) {
        super(extensionFeature);
    }

    @Override
    public String getName() {
        return "Decorator value";
    }

    @Override
    public String getDescription() {
        return "Value.";
    }

    @Override
    public Set<ProjectEntityType> getSupportedEntityTypes() {
        return EnumSet.allOf(ProjectEntityType.class);
    }

    @Override
    public boolean canEdit(ProjectEntity entity, SecurityService securityService) {
        return securityService.isProjectFunctionGranted(entity, ProjectEdit.class);
    }

    @Override
    public boolean canView(ProjectEntity entity, SecurityService securityService) {
        return true;
    }

    @Override
    public Form getEditionForm(ProjectEntity entity, TestDecorationData value) {
        return Form.create()
                .with(
                        Text.of("value")
                                .label("Value")
                                .value(value != null ? value.getValue() : "")
                )
                .with(
                        YesNo.of("valid")
                                .label("Valid")
                                .value(value != null && value.isValid())
                )
                ;
    }

    @Override
    public TestDecorationData fromClient(JsonNode node) {
        return fromStorage(node);
    }

    @Override
    public TestDecorationData fromStorage(JsonNode node) {
        return parse(node, TestDecorationData.class);
    }

    @Override
    public TestDecorationData replaceValue(TestDecorationData value, Function<String, String> replacementFunction) {
        return new TestDecorationData(
                replacementFunction.apply(value.getValue()),
                value.isValid()
        );
    }
}

File: ontrack-extension-api/src/testFixtures/kotlin/net/nemerosa/ontrack/extension/api/support/TestDecoratorPropertyType.kt
package net.nemerosa.ontrack.extension.api.support

import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.support.AbstractPropertyType
import net.nemerosa.ontrack.model.json.schema.JsonType
import net.nemerosa.ontrack.model.json.schema.JsonTypeBuilder
import net.nemerosa.ontrack.model.json.schema.toType
import net.nemerosa.ontrack.model.security.ProjectEdit
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import org.springframework.stereotype.Component
import java.util.*

@Component
class TestDecoratorPropertyType(
    extensionFeature: TestExtensionFeature
) : AbstractPropertyType<TestDecorationData>(extensionFeature) {

    override val name: String = "Decorator value"

    override val description: String = "Value."

    override val supportedEntityTypes: Set<ProjectEntityType> = EnumSet.allOf(ProjectEntityType::class.java)

    override fun createConfigJsonType(jsonTypeBuilder: JsonTypeBuilder): JsonType =
        jsonTypeBuilder.toType(TestDecorationData::class)

    override fun canEdit(entity: ProjectEntity, securityService: SecurityService): Boolean =
        securityService.isProjectFunctionGranted(entity, ProjectEdit::class.java)

    override fun canView(entity: ProjectEntity, securityService: SecurityService): Boolean = true

    override fun fromClient(node: JsonNode): TestDecorationData {
        return fromStorage(node)
    }

    override fun fromStorage(node: JsonNode): TestDecorationData {
        return parse(node, TestDecorationData::class)
    }

    override fun replaceValue(
        value: TestDecorationData,
        replacementFunction: (String) -> String
    ): TestDecorationData {
        return TestDecorationData(
            replacementFunction(value.value),
            value.isValid
        )
    }
}
